src/back/wgsl/writer.rs

```diff
@@ -819,56 +877,136 @@ impl<W: Write> Writer<W> {
                 image,
                 sampler,
                 coordinate,
-                array_index: _,
-                offset: _,
+                array_index,
+                offset,
                 level,
-                depth_ref: _,
+                depth_ref,
             } => {
-                // TODO: other texture functions
-                // TODO: comments
-                let fun_name = match level {
-                    crate::SampleLevel::Auto => "textureSample",
-                    _ => {
-                        return Err(Error::Unimplemented(format!(
-                            "expression_imagesample_level {:?}",
-                            level
-                        )));
+                let texture_function = if let Some(depth_ref) = depth_ref {
+                    TextureFunction::SampleCompare(depth_ref)
+                } else {
+                    match level {
+                        crate::SampleLevel::Auto | crate::SampleLevel::Zero => {
+                            TextureFunction::Sample
+                        }
+                        crate::SampleLevel::Exact(expr) => TextureFunction::SampleLevel(expr),
+                        crate::SampleLevel::Bias(expr) => TextureFunction::SampleBias(expr),
+                        crate::SampleLevel::Gradient { x, y } => {
+                            TextureFunction::SampleGrad { x, y }
+                        }
                     }
                 };
-                write!(self.out, "{}(", fun_name)?;
+
+                write!(self.out, "{}(", texture_function)?;
                 self.write_expr(module, image, func_ctx)?;
                 write!(self.out, ", ")?;
                 self.write_expr(module, sampler, func_ctx)?;
                 write!(self.out, ", ")?;
                 self.write_expr(module, coordinate, func_ctx)?;
+
+                if let Some(array_index) = array_index {
+                    write!(self.out, ", ")?;
+                    self.write_expr(module, array_index, func_ctx)?;
+                }
+
+                if let TextureFunction::SampleBias(expr) = texture_function {
```

Comment:
let's take this out of `level` instead: `if let crate::SampleLevel::Bias(expr) = level`.
Same for other cases